### ISSUE TITLE
iwyu should test the include s

### ISSUE THREAD

**Issue by vsapsai** [include-what-you-use/include-what-you-use#109]
Originally reported on Google Code with ID 109
```
Let's take the following file:
---
#include <glib.h>
#include <stdio.h>

int main() {
  char* f = g_strdup_printf("%s", "test");
  printf("%s\n", f);
  g_free(f);
  return 0;
}
---

iwyu recommends:
$ iwyu `pkg-config --cflags glib-2.0` foo.c 
[...]

The full include-list for foo.c:
#include <stdio.h>                      // for printf
#include "glib/gmem.h"                  // for g_free
#include "glib/gstrfuncs.h"             // for g_strdup_printf

If I rewrite the file as suggested, I get:

---
#include <stdio.h>                      // for printf
#include "glib/gmem.h"                  // for g_free
#include "glib/gstrfuncs.h"             // for g_strdup_printf

  int main() {
    char* f = g_strdup_printf("%s", "test");
    printf("%s\n", f);
    g_free(f);
    return 0;
  }
---

which will fail to build with:
---
In file included from foo.c:2:
/usr/include/glib-2.0/glib/gmem.h:31:2: error: "Only <glib.h> can be included
      directly."
#error "Only <glib.h> can be included directly."
 ^
---

A way for iwyu to check if it works would be to include the header in a temporary file
and build it.

```

Reported by `sylvestre.ledru` on 2013-09-09 07:04:25

**Comment by vsapsai** [comment]
Fixed by PR #288.

**Comment by vsapsai** [comment]
Turned out I was wrong about `#include_next`. It worked on Mac OS X because macros are reported in `IwyuPreprocessorInfo::HandlePreprocessingDone` after we exited from the file. So we don't see a bunch of macros and don't try to map limits.h to limits.h. Looks like handling macros defined by includers should be moved later.

I have an idea to solve limits.h by ignoring `#include_next` includes. I can distinguish `#include_next` includes but something is still not working on Ubuntu.

Another issue is `stddef.h`. `/usr/include/_G_config.h` does something like

```c++
#define __need_NULL
#include <stddef.h>
```

and then `../lib/clang/3.9.0/include/stddef.h` uses this macro to decide if it needs to define NULL. It completely follows the pattern of used macro being defined by includer. But I don't want to map `stddef.h` to `_G_config.h`. Still thinking how to fix this issue. So far, reusing the existing file mapping approach fails in many cases.

**Comment by vsapsai** [comment]
I had ideas how to detect files like `../lib/clang/3.9.0/include/stddef.h` but they are too complex and brittle. My last idea is to check with `IncludePicker::IsPublic(const clang::FileEntry* file)` and avoid marking `file` as private if it is public. It looks like overly direct workaround but it is simple and will cover all cases when heuristic conflicts with explicit mappings.

Quick test on Ubuntu shows it works, will test more and submit a new pull request.

**Comment by kimgr** [comment]
Cool, that makes sense to me. If the header is public, it can't have a hard dependency on a config symbol. And conversely, an x-macro header *always* has a hard dependency. 

Just a thought: why are you doing this analysis for system headers?

**Comment by vsapsai** [comment]
I need to do analysis for system headers to cover glibc. Of course, I can try to exclude Clang internal headers (those that are located in `../lib/clang/3.9.0/include/`). But I need to exclude both file defining a macro and file using a macro. And it won't work for non-internal headers. Given all these factors, I decided that `IncludePicker::IsPublic` is pretty good choice.

**Comment by kimgr** [comment]
OK, when I take a step back, I think I understand the motivation -- the analysis is done to (possibly) produce a mapping, right? And the mappings apply to system headers just as much as user headers -- if you can get this to work really well, we might be able to throw away some builtin mappings.

I'd love to see a new PR on this!

**Comment by kimgr** [comment]
This appears to be fixed by PR #328
